### PR TITLE
Add lint check for DBotRole in scripts with elevated permissions

### PR DESCRIPTION
**Pull request by yaakovpraisler**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://github.com/demisto/etc/issues/28672

## Description
Added lint enforcing for DBotRole in scripts that require elevated permissions.

## Must have
- [x] Tests
- [ ] Documentation

### CODE REVIEW

**Comment by yaakovpraisler** [comment]
In this XSOAR lint validate we check the following use cases:

- If there is in the code a call to the 'demisto.executeCommand' function, with the built in 'DeleteContext' function as an argument. For example `demisto.executeCommand('DeleteContext')`
- If there is in the code a call to the 'demisto.executeCommand' function, with the built in 'getIncidents' function as an argument. For example `demisto.executeCommand('getIncidents')`
- If there is in the code a call to the 'demisto.executeCommand' function, with the built in 'setIncident' function as an argument, and also an incident id as an argument. For example `demisto.executeCommand('setIncident', {'id': incident_id})`

In all three cases above we will only return an error if 'runas: DBotRole' does not exist in the script yml file.

File: demisto_sdk/commands/lint/resources/pylint_plugins/base_checker.py [hunk 1]
```diff
@@ -33,7 +33,10 @@
               "unimplemented-test-module",
               "test-module command is not implemented in the python file, it is essential for every"
               " integration. Please add it to your code. For more information see: "
-              "https://xsoar.pan.dev/docs/integrations/code-conventions#test-module")
+              "https://xsoar.pan.dev/docs/integrations/code-conventions#test-module"),
+    "E9012": ("Your script require elevated permissions. Please add the `runas: DBotRole` to the yml file.",
```
Comment (roysagi) [inline demisto_sdk/commands/lint/resources/pylint_plugins/base_checker.py:37]:
```suggestion
    "E9012": ("Your script requires elevated permissions. Please add the `runas: DBotRole` to the yml file.",
```

File: demisto_sdk/commands/lint/resources/pylint_plugins/base_checker.py [hunk 2]
```diff
@@ -33,7 +33,10 @@
...
+              "missing-permission",
+              "Your script require elevated permissions. Please add the `runas: DBotRole` to the yml file.")
```
Comment (roysagi) [inline demisto_sdk/commands/lint/resources/pylint_plugins/base_checker.py:39]:
```suggestion
              "Your script requires elevated permissions. Please add the `runas: DBotRole` to the yml file.")
```

File: demisto_sdk/commands/lint/resources/pylint_plugins/base_checker.py [hunk 3]
```diff
@@ -122,6 +127,28 @@ def _quit_checker(self, node):
         except Exception:
             pass
 
+    def _execute_command_checker(self, node):
```
Comment (roysagi) [inline demisto_sdk/commands/lint/resources/pylint_plugins/base_checker.py:130]: please add a docstring and possibly some deev comments in the method

File: demisto_sdk/commands/lint/resources/pylint_plugins/base_checker.py [hunk 4]
```diff
@@ -122,6 +127,28 @@ def _quit_checker(self, node):
...
+        if not self.is_script or self.runas == 'DBotRole':
+            return
+
+        try:
+            if node.func.expr.name == 'demisto' and node.func.attrname == 'executeCommand':
+
+                if node.args[0].value == 'getIncidents' or node.args[0].value == 'DeleteContext':
```
Comment (roysagi) [inline demisto_sdk/commands/lint/resources/pylint_plugins/base_checker.py:137]:
```suggestion
                if node.args[0].value in ['getIncidents', 'DeleteContext']:
```

File: demisto_sdk/commands/lint/tests/test_pylint_plugin/base_checker_test.py [hunk 5]
```diff
@@ -739,3 +739,143 @@ def test_indicators_doesnt_exist(self):
...
+    CHECKER_CLASS = base_checker.CustomBaseChecker
+
+    def test_only_execute_command_exists(self):
+        """
+        Given:
+            - String of a code part which is being examined by pylint plugin.
+        When:
+            - executeCommand function exists in the code without any build in command.
+        Then:
+            - Ensure that there is no errors, Check that there is no error message.
```
Comment (roysagi) [inline demisto_sdk/commands/lint/tests/test_pylint_plugin/base_checker_test.py:757]:
```suggestion
            - Ensure that there are no errors, check that there is no error message.
```

File: demisto_sdk/commands/lint/tests/test_pylint_plugin/base_checker_test.py [hunk 6]
```diff
@@ -739,3 +739,143 @@ def test_indicators_doesnt_exist(self):
...
+    Class which tests the functionality of executeCommand checker.
+    """
+    CHECKER_CLASS = base_checker.CustomBaseChecker
+
+    def test_only_execute_command_exists(self):
+        """
+        Given:
+            - String of a code part which is being examined by pylint plugin.
+        When:
+            - executeCommand function exists in the code without any build in command.
```
Comment (roysagi) [inline demisto_sdk/commands/lint/tests/test_pylint_plugin/base_checker_test.py:755]:
```suggestion
            - executeCommand function exists in the code without any build-in command.
```

File: demisto_sdk/commands/lint/tests/test_pylint_plugin/base_checker_test.py [hunk 7]
```diff
@@ -739,3 +739,143 @@ def test_indicators_doesnt_exist(self):
...
+        assert node_b is not None
+        with self.assertNoMessages():
+            self.checker.visit_call(node_b)
+
+    def test_execute_command_with_delete_context_exists(self):
+        """
+        Given:
+            - String of a code part which is being examined by pylint plugin.
+        When:
+            - executeCommand function exists in the code with the 'DeleteContext' build in command.
```
Comment (roysagi) [inline demisto_sdk/commands/lint/tests/test_pylint_plugin/base_checker_test.py:773]:
```suggestion
            - executeCommand function exists in the code with the 'DeleteContext' build-in command.
```

File: demisto_sdk/commands/lint/tests/test_pylint_plugin/base_checker_test.py [hunk 8]
```diff
@@ -739,3 +739,143 @@ def test_indicators_doesnt_exist(self):
...
+        self.checker.is_script = True
+        _, node_b, node_c = astroid.extract_node("""
+            def test_function(): #@
+                args = {'name': 'incident_name'} #@
+                demisto.executeCommand('setIncident', args) #@
+        """)
+        assert node_b is not None and node_c is not None
+        with self.assertNoMessages():
+            self.checker.visit_call(node_b)
+            self.checker.visit_call(node_c)
```
Comment (roysagi) [inline demisto_sdk/commands/lint/tests/test_pylint_plugin/base_checker_test.py:881]: let's add a test to see this test is skipped if runas is set.

File: demisto_sdk/commands/lint/resources/pylint_plugins/base_checker.py (hunk 1 above)
Comment (roysagi) [reply demisto_sdk/commands/lint/resources/pylint_plugins/base_checker.py:37]: should this be an error or warning @Itay4?

**Review by roysagi — CHANGES_REQUESTED** [review]
@yaakovpraisler - seems good, do see my comments about making the documentation clearer also please add a changelog entry.

File: demisto_sdk/commands/lint/resources/pylint_plugins/base_checker.py (hunk 1 above)
Comment (yaakovpraisler) [reply demisto_sdk/commands/lint/resources/pylint_plugins/base_checker.py]: done.

File: demisto_sdk/commands/lint/resources/pylint_plugins/base_checker.py (hunk 2 above)
Comment (yaakovpraisler) [reply demisto_sdk/commands/lint/resources/pylint_plugins/base_checker.py]: done.

File: demisto_sdk/commands/lint/resources/pylint_plugins/base_checker.py (hunk 3 above)
Comment (yaakovpraisler) [reply demisto_sdk/commands/lint/resources/pylint_plugins/base_checker.py:131]: done.

File: demisto_sdk/commands/lint/resources/pylint_plugins/base_checker.py (hunk 4 above)
Comment (yaakovpraisler) [reply demisto_sdk/commands/lint/resources/pylint_plugins/base_checker.py]: done.
Also added here the additional command, and put them in the constants file.

File: demisto_sdk/commands/lint/tests/test_pylint_plugin/base_checker_test.py (hunk 6 above)
Comment (yaakovpraisler) [reply demisto_sdk/commands/lint/tests/test_pylint_plugin/base_checker_test.py]: done.

File: demisto_sdk/commands/lint/tests/test_pylint_plugin/base_checker_test.py (hunk 5 above)
Comment (yaakovpraisler) [reply demisto_sdk/commands/lint/tests/test_pylint_plugin/base_checker_test.py]: done.

File: demisto_sdk/commands/lint/tests/test_pylint_plugin/base_checker_test.py (hunk 7 above)
Comment (yaakovpraisler) [reply demisto_sdk/commands/lint/tests/test_pylint_plugin/base_checker_test.py]: done.

File: demisto_sdk/commands/lint/tests/test_pylint_plugin/base_checker_test.py (hunk 8 above)
Comment (yaakovpraisler) [reply demisto_sdk/commands/lint/tests/test_pylint_plugin/base_checker_test.py:862]: done.

File: demisto_sdk/commands/lint/resources/pylint_plugins/base_checker.py [hunk 9]
```diff
@@ -122,6 +128,40 @@ def _quit_checker(self, node):
...
 
+    def _execute_command_checker(self, node):
+        """
+        Checks if there is a call to the executeCommand function with build-in commands, and returns an error message
+        if 'runas: DBotRole' not set in script.
+
+        Args:
+            node: A astroid node.
+        """
+        # If it's not a script or 'runas' is already set to 'DBotRole'.
```
Comment (roysagi) [inline demisto_sdk/commands/lint/resources/pylint_plugins/base_checker.py:139]:
```suggestion
        # If it's not a script or 'runas' is already set to 'DBotRole' simply return.
```

**Review by roysagi — APPROVED** [review]

File: demisto_sdk/tests/integration_tests/xsoar_linter_integration_test.py [hunk 10]
```diff
@@ -33,9 +33,13 @@
...
         " integration. Please add it to your code. For more information see: "
         "https://xsoar.pan.dev/docs/integrations/code-conventions#test-module"
-    ], []),
+    ], [], False),
 
     # -------------------- For Invalid file with support level base and long running False -----------------------
+
+    (Path(f"{XSOAR_LINTER_PY3_INVALID}"), 3.8, 'base', False, 1, [
+        "Your script requires elevated permissions. Please add the `runas: DBotRole` to the yml file."
+    ], [], True),
```
Comment (orhovy) [inline demisto_sdk/tests/integration_tests/xsoar_linter_integration_test.py:42]:
```suggestion
    ], [], True),
    
    # -------------------- For Invalid file with support level base and long running False -----------------------

```

File: demisto_sdk/tests/integration_tests/xsoar_linter_integration_test.py [hunk 11]
```diff
@@ -33,9 +33,13 @@
         "test-module command is not implemented in the python file, it is essential for every"
         " integration. Please add it to your code. For more information see: "
         "https://xsoar.pan.dev/docs/integrations/code-conventions#test-module"
-    ], []),
+    ], [], False),
 
     # -------------------- For Invalid file with support level base and long running False -----------------------
```
Comment (orhovy) [inline demisto_sdk/tests/integration_tests/xsoar_linter_integration_test.py:38]:
```suggestion
    # -------------------- For Invalid file with support level base and is a script -----------------------
```

File: demisto_sdk/tests/integration_tests/xsoar_linter_integration_test.py (hunk 10 above)
Comment (orhovy) [inline demisto_sdk/tests/integration_tests/xsoar_linter_integration_test.py:42]:
```suggestion
    ], [], True),
    
    # -------------------- For Invalid file with support level base and long running False -----------------------

```

File: demisto_sdk/tests/integration_tests/xsoar_linter_integration_test.py [hunk 12]
```diff
@@ -33,9 +33,16 @@
...
+    ], [], False),
 
+    # -------------------- For Invalid file with support level base and is a script -----------------------
+
+    (Path(f"{XSOAR_LINTER_PY3_INVALID}"), 3.8, 'base', False, 1, [
+        "Your script requires elevated permissions. Please add the `runas: DBotRole` to the yml file."
+    ], [], True),
+    
     # -------------------- For Invalid file with support level base and long running False -----------------------
+
```
Comment (orhovy) [inline demisto_sdk/tests/integration_tests/xsoar_linter_integration_test.py:45]:
```suggestion
```

File: demisto_sdk/tests/integration_tests/xsoar_linter_integration_test.py [hunk 13]
```diff
@@ -33,8 +33,14 @@
         "test-module command is not implemented in the python file, it is essential for every"
         " integration. Please add it to your code. For more information see: "
         "https://xsoar.pan.dev/docs/integrations/code-conventions#test-module"
-    ], []),
+    ], [], False),
 
+    # -------------------- For Invalid file with support level base and is a script -----------------------
+
```
Comment (orhovy) [inline demisto_sdk/tests/integration_tests/xsoar_linter_integration_test.py:39]:
```suggestion
```

**Review by orhovy — APPROVED** [review]

**Comment by yaakovpraisler** [comment]
Waiting for the content updates in [this PR](https://github.com/demisto/content/pull/13455).

**Pull request merged**